### PR TITLE
Add test for BibTeX key validation

### PR DESCRIPTION
**Pull request by wupr**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -128,6 +135,10 @@ enum Command {
         /// Write output to file.
         #[arg(short, long)]
         out: Option<PathBuf>,
+        /// The output format.
+        /// Select `null` to suppress BibTeX output and disable checks on citation keys.
```
Comment (alexrutar) [inline src/main.rs:139]: I think you need to leave a space after the first line, or it creates an overflowed help message when you run `autobib help get`.

File: src/main.rs [hunk 2]
```diff
@@ -538,10 +558,13 @@ fn run_cli(cli: Cli) -> Result<()> {
...
                 &mut record_db,
                 &client,
+                format,
                 ignore_null,
             );
 
-            output_entries(out.as_ref(), valid_entries)?;
+            if format != OutputFormat::Null {
+                output_entries(out.as_ref(), valid_entries)?;
+            }
```
Comment (alexrutar) [inline src/main.rs:567]: I would probably prefer if the `write_entries` / `output_entries` signature were modified to include the `OutputFormat` as an argument, to make it easier to implement new formats in the future.

File: src/main.rs [hunk 3]
```diff
@@ -730,19 +754,26 @@ fn retrieve_and_validate_single_entry(
...
+                } = record;
+                let entry = validate_bibtex_entry(key, data, &row).map(|entry| (entry, canonical));
+                row.commit()?;
+                Ok(entry)
+            }
+            OutputFormat::Null => {
+                row.commit()?;
+                Ok(None)
+            }
+        },
```
Comment (alexrutar) [inline src/main.rs:776]: I don't really think this optimization is worthwhile, it increases the complexity of this implementation for a rather uncommon function call.

Comment (wupr) [reply src/main.rs:776]: What would you suggest we do? `validate_bibtex_entry` just handles the key validation, so I think it makes sense to skip it altogether.

Comment (alexrutar) [reply src/main.rs:776]: Oh I see, of course when the `Entry` is generated, the error is already cause.

File: src/entry.rs [hunk 4]
```diff
@@ -9,37 +9,55 @@ use serde_bibtex::{de::Deserializer, to_string_unchecked, validate::is_entry_key
...
-    pub key: String,
-    pub record_data: D,
+pub struct BibtexKey {
+    key: String,
 }
 
-impl<D: EntryData> Entry<D> {
+impl BibtexKey {
     /// Create a new entry data with the provided key.
     ///
```
Comment (alexrutar) [inline src/entry.rs:22]: Typo: should be updated

File: src/entry.rs [hunk 5]
```diff
@@ -9,37 +9,55 @@ use serde_bibtex::{de::Deserializer, to_string_unchecked, validate::is_entry_key
...
-    ///
-    /// # Safety
-    /// The caller is required to guarantee that the key does not contain any characters which are
-    /// invalid BibTeX entry key characters, as accepted by the [`serde_bibtex::validate::is_entry_key`] method.
-    pub(crate) fn new_unchecked(key: String, record_data: D) -> Self {
-        Self { key, record_data }
+/// A single regular entry in a BibTeX bibliography.
+#[derive(Debug, PartialEq)]
+pub struct Entry<D: EntryData> {
+    pub key: String,
```
Comment (alexrutar) [inline src/entry.rs:50]: Should this rather be a `BibtexKey`?

Comment (alexrutar) [reply src/entry.rs:50]: If this is changed, I guess the `Serializer` impl will break (but is easy to fix), but hopefully nothing else.

Comment (alexrutar) [reply src/entry.rs:50]: Also, if this is changed, for convenience, we should implement `From<BibtexKey> for String`

Comment (alexrutar) [reply src/entry.rs:50]: If we go down this route, it should also implement `AsRef<str>` which is useful to fix the errors in `main.rs`. In which case I would even `BibtexKey` being a "newtype"-style struct.

Comment (alexrutar) [reply src/entry.rs:50]: Oh I just realized I literally implemented this once already: https://docs.rs/serde_bibtex/latest/serde_bibtex/token/struct.EntryKey.html

But the API is quite bad, if we want to merge these changes in, then I should update upstream...

Comment (wupr) [reply src/entry.rs]: I've added the suggested implementations. Happy to replace with `EntryKey` if you want to update upstream.

**Comment by alexrutar** [comment]
Let me fix upstream and see how it goes.

**Comment by alexrutar** [comment]
I'm most of the way through implementing this upstream; unfortunately, a pre-requisite was to fix all of the horrifying error handling in that crate which will take a bit of time.

**Pull request merged**